notification/slack/usergroupsender.go

```diff
@@ -0,0 +1,169 @@
+package slack
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	"github.com/slack-go/slack"
+	"github.com/target/goalert/config"
+	"github.com/target/goalert/notification"
+	"github.com/target/goalert/permission"
+	"github.com/target/goalert/user"
+	"github.com/target/goalert/util/log"
+)
+
+// UserGroupSender processes on-call notifications by updating the members of a Slack user group.
+type UserGroupSender struct {
+	*ChannelSender
+}
+
+var _ notification.Sender = (*UserGroupSender)(nil)
+
+// UserGroupSender returns a new UserGroupSender wrapping the given ChannelSender.
+func (s *ChannelSender) UserGroupSender() *UserGroupSender {
+	return &UserGroupSender{s}
+}
+
+// Send implements notification.Sender.
+func (s *UserGroupSender) Send(ctx context.Context, msg notification.Message) (*notification.SentMessage, error) {
+	err := permission.LimitCheckAny(ctx, permission.User, permission.System)
+	if err != nil {
+		return nil, err
+	}
+
+	t, ok := msg.(notification.ScheduleOnCallUsers)
+	if !ok {
+		return nil, errors.Errorf("unsupported message type: %T", msg)
+	}
+
+	if t.Dest.Type != notification.DestTypeSlackUG {
+		return nil, errors.Errorf("unsupported destination type: %s", t.Dest.Type.String())
+	}
+
+	teamID, err := s.TeamID(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("lookup team ID: %w", err)
+	}
+
+	var userIDs []string
+	for _, u := range t.Users {
+		userIDs = append(userIDs, u.ID)
+	}
+
+	userSlackIDs := make(map[string]string, len(t.Users))
+	err = s.cfg.UserStore.AuthSubjectsFunc(ctx, "slack:"+teamID, userIDs, func(sub user.AuthSubject) error {
```

Comment:
I disagree, `Sprintf` has its uses, but it doesn't need to be used in every possible place, we should instead use the right tool for the job. In this case, I don't see how a call into `Sprintf` makes the code more readable or concise.

Reply:
The choice between `fmt.Sprintf("test/%s", val)` and `"test/" + val` should be primarily based on performance, readability, and consistency. For simple string concatenations, the `+` operator is generally more efficient and easier to read. If you have more complex formatting requirements, `fmt.Sprintf` can lead to cleaner and more readable code.

Reply:
I believe there might be some misunderstanding regarding the usage of `%s`. It's important to note that `%s` is essentially performing direct string concatenation and does not offer any escaping or protection in any circumstance across languages.

My concern is that spreading inaccurate information (i.e., suggesting that `Sprintf` with `%s` provides any level of security) could be counterproductive. While I appreciate your dedication to secure coding practices, it's essential that we focus on accurate information and context-specific solutions.

If there's a pragmatic reason for your suggestion, I'm more than willing to discuss that further. However, it would be concerning if any training material implies that using a formatting function with raw strings inherently makes the code more secure. Our primary goal should be to write effective code while ensuring we share accurate and helpful information with our colleagues.